### ISSUE TITLE
Logging MultiIndex Dataframes Fails

### ISSUE THREAD

**Issue by Lazea** [capitalone/rubicon#47]
**Describe the bug**
It appears that internally, rubicon's `.log_dataframe()` converts pandas dataframes to dask dataframes regardless of the situation. This can cause issues in scenarios where dask might not support certain dataframe features such as multiindex dataframes. 

**Steps/Code to reproduce bug**
```python
import pandas as pd
from rubicon.client import Rubicon
# Create sample data
df = pd.DataFrame([[0,1,'a'],[1,1,'b'],[2,2,'c'],[3,2,'d']], columns=['a', 'b', 'c'])
df = df.set_index(['b', 'a']) # Set multiindex
df
     c
b a   
1 0  a
  1  b
2 2  c
  3  d

# Log dataframe to rubicon
rubicon = Rubicon(persistence="memory")
project = rubicon.get_or_create_project("test")
exp = project.log_experiment('test_exp')
exp.log_dataframe(df)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ouz343/miniconda3/envs/lustr/lib/python3.8/site-packages/rubicon/client/mixin.py", line 251, in log_dataframe
    self.repository.create_dataframe(dataframe, df, project_name, experiment_id=experiment_id)
  File "/Users/ouz343/miniconda3/envs/lustr/lib/python3.8/site-packages/rubicon/repository/base.py", line 426, in create_dataframe
    data = self._convert_to_dask_dataframe(data)
  File "/Users/ouz343/miniconda3/envs/lustr/lib/python3.8/site-packages/rubicon/repository/base.py", line 396, in _convert_to_dask_dataframe
    return dd.from_pandas(df, npartitions=1)
  File "/Users/ouz343/miniconda3/envs/lustr/lib/python3.8/site-packages/dask/dataframe/io/io.py", line 202, in from_pandas
    raise NotImplementedError("Dask does not support MultiIndex Dataframes.")
NotImplementedError: Dask does not support MultiIndex Dataframes.
```

**Additional context**
Not familiar as to why pandas dataframes need to be converted to dask dataframes every time during logging but the solution would revolve around avoiding conversion to dask since dask in this case does not support multiindex.

**Comment by ryanSoley** [comment]
yeah I don't see any reason we should be doing this [`_convert_to_dask_dataframe`](https://github.com/capitalone/rubicon/blob/main/rubicon/repository/base.py#L391) step in each dataframe write

we should just remove it and make sure [`_persist_dataframe`](https://github.com/capitalone/rubicon/blob/main/rubicon/repository/base.py#L398) works for `dask` and `pandas` dataframes. (it should, just wanna double check the `engine="pyarrow"` kwarg is acceptable for both)

we'll also have to modify [`_read_dataframe`](https://github.com/capitalone/rubicon/blob/main/rubicon/repository/base.py#L402) to fall back to a `pandas` read if the `dask` read fails. may even wanna expose an `as_type` kwarg up to the client on dataframe reads to specify how you want to read the data since we can't infer it without storing extra info, and that seems unnecessary here

**Comment by joe-wolfe21** [comment]
> we'll also have to modify _read_dataframe to fall back to a pandas read if the dask read fails. may even wanna expose an as_type ("pandas" or "dask") kwarg up to the client on dataframe reads to specify how you want to read the data since we can't infer it without storing extra info, and that seems unnecessary here

I think we'll have to either expose `as_type` on the read or store the type as additional meta because from my testing it looks like `dd.read_parquet(path, engine="pyarrow")` doesn't actually fail on the multi_index read, it just returns an empty df (which we could check for but the other ideas seem better)

Any preferences? I could see not wanting to store extra meta, but it would make it automatic for users rather than having to specify on the client read

**Comment by joe-wolfe21** [comment]
Using `df.to_parquet(path, engine="pyarrow")` to persist either a pandas or dask df and using `dd.read_parquet(path, engine="pyarrow")` to return a computed pandas df seems to work for all cases

```py
>>> import pandas as pd
>>> from rubicon.client import Rubicon
>>>
>>> df = pd.DataFrame([[0,1,'a'],[1,1,'b'],[2,2,'c'],[3,2,'d']], columns=['a', 'b', 'c'])
>>> df = df.set_index(['b', 'a']) # Set multiindex
>>> df
     c
b a
1 0  a
  1  b
2 2  c
  3  d
>>>
>>> rubicon = Rubicon(persistence="memory")
>>> project = rubicon.get_or_create_project("test")
>>> exp = project.log_experiment('test_exp')
>>> exp.log_dataframe(df)
<rubicon.client.dataframe.Dataframe object at 0x7fda79ec7e20>
>>> exp.dataframes()[0].data
     c
b a
1 0  a
  1  b
2 2  c
  3  d
>>> isinstance(exp.dataframes()[0].data, pd.DataFrame)
True
```

I've included tests for pandas, multiindex pandas, and dask. If this looks good, I'll apply to the async methods as well

**Comment by joe-wolfe21** [comment]
@Lazea thanks for raising this issue -- released in version `0.1.5`

**Comment by Lazea** [comment]
Thanks for getting this together.
